Sources/_MatchingEngine/Regex/Parse/LexicalAnalysis.swift

```diff
@@ -317,56 +329,56 @@ extension Source {
   }
 
-  /// Consume a range
+  /// Try to consume a range, returning `nil` if unsuccessful.
   ///
   ///     Range       -> ',' <Int> | <Int> ',' <Int>? | <Int>
   ///                  | ExpRange
   ///     ExpRange    -> '..<' <Int> | '...' <Int>
   ///                  | <Int> '..<' <Int> | <Int> '...' <Int>?
-  mutating func expectRange() throws -> Located<Quant.Amount> {
+  mutating func lexRange() throws -> Located<Quant.Amount>? {
     try recordLoc { src in
-      // TODO: lex positive numbers, more specifically...
-
-      let lowerOpt = try src.lexNumber()
-
-      // ',' or '...' or '..<' or nothing
-      let closedRange: Bool?
-      if src.tryEat(",") {
-        closedRange = true
-      } else if src.experimentalRanges && src.tryEat(".") {
-        try src.expect(".")
-        if src.tryEat(".") {
+      try src.tryEating { src in
+        let lowerOpt = try src.lexNumber()
+
+        // ',' or '...' or '..<' or nothing
+        // TODO: We ought to try and consume whitespace here and emit a
+        // diagnostic for the user warning them that it would cause the range to
+        // be treated as literal.
+        let closedRange: Bool?
+        if src.tryEat(",") {
           closedRange = true
+        } else if src.experimentalRanges && src.tryEat(".") {
+          try src.expect(".")
+          if src.tryEat(".") {
+            closedRange = true
+          } else {
+            try src.expect("<")
+            closedRange = false
+          }
         } else {
-          try src.expect("<")
-          closedRange = false
+          closedRange = nil
         }
-      } else {
-        closedRange = nil
-      }
-      // FIXME: wait, why `try!` ?
-      let upperOpt: Located<Int>?
-      if let u = try! src.lexNumber() {
-        upperOpt = (closedRange == true) ? u : Located(u.value-1, u.location)
-      } else {
-        upperOpt = nil
-      }
 
-      switch (lowerOpt, closedRange, upperOpt) {
-      case let (l?, nil, nil):
-        return .exactly(l)
-      case let (l?, true, nil):
-        return .nOrMore(l)
-      case let (nil, _, u?):
-        return .upToN(u)
-      case let (l?, _, u?):
-        // FIXME: source location tracking
-        return .range(l, u)
-
-      case let (nil, nil, u) where u != nil:
-        fatalError("Not possible")
-      default:
-        throw ParseError.misc("Invalid range")
+        let upperOpt = try src.lexNumber()?.map { upper in
+          // If we have an open range, the upper bound should be adjusted down.
+          closedRange == true ? upper : upper - 1
+        }
+
+        switch (lowerOpt, closedRange, upperOpt) {
+        case let (l?, nil, nil):
+          return .exactly(l)
+        case let (l?, true, nil):
+          return .nOrMore(l)
+        case let (nil, _?, u?):
+          return .upToN(u)
+        case let (l?, _?, u?):
+          return .range(l, u)
+
+        case (nil, nil, _?):
+          fatalError("Didn't lex lower bound, but lexed upper bound?")
```

Comment:
At some point, we should do a pass over the codebase and convert unreachable fatalErrors to unreachable throws, differentiating them from diagnostics.